### ISSUE TITLE
Adding object files to the test executable

### ISSUE THREAD

**Issue by fbrier** [ptomulik/scons-tool-cxxtest#4]
My attempt to have the cxxtest link with a set of .obj files seems to have bumped into your [nice comment](https://github.com/ptomulik/scons-tool-cxxtest/blob/master/__init__.py#L211) about not being able to use the CXXTESTxxx variables. I thought I might be able to append a list of .obj(s) to the CXXTESTLINKFLAGS, but it is ignored. Thoughts?

The list of .obj(s) get bundled into a shared library, but I don't want to have to expose C++ classes that won't be used by the system loading the shared library. My thought was to link them directly into the cxxtest executable/library. Thank you.

**Comment by ptomulik** [comment]
I haven't looked at this project for years, so it's hard to answer quickly. FWIR, ``$[CXXTEST]LINKFLAGS`` is not a variable dedicated for providing additional sources/objects. It's a bad practice.

You may try to compile your ``.obj``s to a library (or libraries) and add the libraries to the executable using ``$CXXTESTLIBS``.

```scons
env.StaticLibrary('foo.a', ['f1.obj', 'f2.obj']);
env.CxxTest('mytest.cpp.t', CXXTESTLIBS=['foo.a']);
```

**Comment by fbrier** [comment]
After working with bdbaddog last night on the SCons Discord channel, the approach of bundling the .obj(s) into a library is the direction I started to go. Thank you for responding. Your SCons Builder was very easy to install and use.

Another need is to mock some of my C++ classes and after looking how to do so with cxxtest I took a look at GoogleMock and since the projects have been merged, GoogleTest. While I did see an old post on using GoogleMock with cxxtest, if I may ask, do you still use and like cxxtest? Or have you moved onto another framework such as GoogleTest? I had selected cxxtest primarily because of your nice integration with SCons. Thank you.

**Comment by ptomulik** [comment]
>  [...] do you still use and like cxxtest? Or have you moved onto another framework such as GoogleTest? I had selected cxxtest primarily because of your nice integration with SCons. Thank you.

Well I haven't programmed anything in C/C++ for long time, and didn't use cxxtest for years, so it's hard to recall. Didn't use GootleTest at all (as well as any of the other possibly existing frameworks).

If I remember correctly, cxxtest provided nothing to support mock objects, it was just said: "[implement it yourself](http://cxxtest.com/guide.html#mock)". CXXTEST provides some macros for mocking global functions.


Years ago I used cxxtest to test my shared libraries in one of my never-completed-projects. If I remember correctly, I compiled code twice to create two versions of each library that needs function mocks to be tested. The first one was a "real" library ``foo.so`` and the other was a "test" library ``foo_tets.so``. It was something similar to this:

```python
# ...
objects = env.SharedObject(sources)

# rlib - "real" libarary
rlib = env.SharedLibrary('foo', objects + ['real.cpp'])

# tlib - library for testing (OpenCL functions replaced with mocks)
tlib = env.SharedLibrary('foo_test', objects + ['mock.cpp'])
# ....
```

The two files ``real.cpp`` and ``mock.cpp`` were

```cpp
// real.cpp
#define CXXTEST_MOCK_REAL_SOURCE_FILE
#include <mock.hpp>
````

```cpp
// mock.cpp
#define CXXTEST_MOCK_TEST_SOURCE_FILE
#include <mock.hpp>
```

and my ``mock.hpp`` was full of CXXTEST mock definitions, here is a fragment:

```cpp
// ...
#include <cxxtest/Mock.h>

/** \cond SHOW_IGNORED_COMPOUNDS */
CXXTEST_MOCK_GLOBAL(cl_int, clGetPlatformIDs,
  (cl_uint num_entries, cl_platform_id* platforms, cl_uint *num_platforms),
  (num_entries, platforms, num_platforms)
);

CXXTEST_MOCK_GLOBAL(cl_int, clGetPlatformInfo,
  ( cl_platform_id platform, cl_platform_info param_name,
    size_t param_value_size, void* param_value, size_t* param_value_size_ret),
  ( platform, param_name, param_value_size, param_value, param_value_size_ret)
);
```

**Comment by fbrier** [comment]
I had read that cxxtest mock documentation and was unhappy about the support for just global functions and not classes. That is why I started to look at GoogleMock. Your response and examples confirm that lack, as well as the approach to compiling two version of a library, one being for test. That is helpful. This project of mine is requiring me to evaluate and assemble a C++ tool chain. Thank you very much for your project and responses. As an aside, I have lots of projects I started and never finished :).
